Adds strict match annotation + updated tests/docs

chrisshino opened the pull request:
## What this PR does / why we need it:
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

This officially closes: https://github.com/kubernetes/ingress-nginx/issues/8330, this patch is needed to satisfy some of the feedbacks received on the original branch that was already merged in.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

chrisshino commented:
https://github.com/kubernetes/ingress-nginx/pull/8434 In this pull request James mentions a few things that would improve this annotation and I implemented them.

1. Pulling out the CN from the client certificates DN (this has been done, meaning no longer do we have to add "CN=" before the regex or string to match.
2.  Having an option to choose between either a strict match or regex match (these have both been implemented here)

rikatz commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -28,7 +28,8 @@ You can add these Kubernetes annotations to specific Ingress objects to customiz
 |[nginx.ingress.kubernetes.io/auth-tls-verify-client](#client-certificate-authentication)|string|
 |[nginx.ingress.kubernetes.io/auth-tls-error-page](#client-certificate-authentication)|string|
 |[nginx.ingress.kubernetes.io/auth-tls-pass-certificate-to-upstream](#client-certificate-authentication)|"true" or "false"|
-|[nginx.ingress.kubernetes.io/auth-tls-match-cn](#client-certificate-authentication)|string|
```
let's not deprecate an annotation. Instead, the old behavior should be called this way, and the new one can have a new name

chrisshino replied:
makes sense! I will keep the original one here as the regex version, and then only add a new one for the strict mode!

rikatz commented on go.mod:
```diff
@@ -15,7 +15,7 @@ require (
 	github.com/moul/pb v0.0.0-20180404114147-54bdd96e6a52
 	github.com/ncabatoff/process-exporter v0.7.10
-	github.com/onsi/ginkgo v1.16.5
+	github.com/onsi/ginkgo v1.16.4
```
why is this being downgraded?

chrisshino replied:
these must have happened after I did a go mod tidy, im not sure why Ill make sure they are the same as before!

rikatz commented on go.mod:
```diff
@@ -40,7 +40,7 @@ require (
 	k8s.io/klog/v2 v2.60.1
 	pault.ag/go/sniff v0.0.0-20200207005214-cf7e4d167732
-	sigs.k8s.io/controller-runtime v0.11.2
+	sigs.k8s.io/controller-runtime v0.10.3
```
why is this being downgraded?

rikatz commented on internal/ingress/annotations/authtls/main_test.go:
```diff
@@ -165,8 +170,11 @@ func TestAnnotations(t *testing.T) {
+	if u.ExactMatchCN != "hello-app" {
+		t.Errorf("expected %v but got %v", "hello-app", u.ExactMatchCN)
+	}
+	if match, _ := regexp.MatchString(u.RegexMatchCN, "hello"); !match {
```
please don't ignore the error here

rikatz commented on internal/ingress/annotations/authtls/main.go:
```diff
@@ -130,9 +130,14 @@ func (a authTLS) Parse(ing *networking.Ingress) (interface{}, error) {
+		config.ExactMatchCN = ""
+	}
+
+	config.RegexMatchCN, err = parser.GetStringAnnotation("auth-tls-regex-match-cn", ing)
```
I want actually that we validate if this is a valid regex per NGINX and not just add it to nginx.conf.

Based on the feedbacks, do we have a common pattern of required strings, like " REGEX = domain.*" or something like that, in a way we can sanitize/remove possible dangerous characters?

If so, let's do this (in both fields actually, we should validate what character set is allowed here)

chrisshino replied:
hmm yea we can do this, I think it will always be like you say something like domain.*, but for example say the person wanted to use a regex match and the CN was = my.example.com and their regex only contained "example", would it fail because it doesn't have a dot after?

Or what if the domain contains numbers? Any suggestions on what a good pattern could look like here? probably something that contains letters, numbers or dots, but no other characters (for example, brackets "<", ">" ?

rikatz reviewed (changes requested):
@chrisshino thanks for the PR!

I've left some comments here, happy to discuss about it.

Thanks

chrisshino commented on internal/ingress/annotations/authtls/main.go:
```diff
@@ -130,9 +130,14 @@ func (a authTLS) Parse(ing *networking.Ingress) (interface{}, error) {
-	config.MatchCN, err = parser.GetStringAnnotation("auth-tls-match-cn", ing)
-	if err != nil || !commonNameRegex.MatchString(config.MatchCN) {
-		config.MatchCN = ""
+	config.ExactMatchCN, err = parser.GetStringAnnotation("auth-tls-exact-match-cn", ing)
```
There should be no priority, it is up to the user which annotation they want to use!

chrisshino replied:
I see what you mean, I think because they are 2 separate annotations then it's hard to prioritize one over the other.
because the original is called auth-tls-match-cn, maybe we can check that one first and if it's not empty then we can exit early. If it is empty we can then check the auth-tls-exact-match-cn

chrisshino replied:
since they are essentially 2 different annotations there isn't necessarily a priority.

chrisshino commented:
@rikatz thank you for the thorough review 🙏 , I believe this should address your comments

chrisshino commented:
hmm so whenever the test runs the go mod is switching those 2 dependency that you called out above to the lower versions?

Then it is failing, not sure why?

iamNoah1 reviewed (commented):
/triage accepted
/priority backlog

iamNoah1 commented:
/kind bug

iamNoah1 commented:
@chrisshino following up on this one. Is it still failing for you, the checks look good for me. If you incorporated all change requests, we could move on here :)

chrisshino commented:
yup everything is working! Thanks for following up on this

rikatz commented on internal/ingress/annotations/authtls/main.go:
```diff
@@ -36,7 +36,7 @@ const (
 var (
 	authVerifyClientRegex = regexp.MustCompile(`on|off|optional|optional_no_ca`)
-	commonNameRegex       = regexp.MustCompile(`CN=`)
+	matchCNSanitizer      = regexp.MustCompile(`^[a-zA-Z0-9\.\-_\(\)\|]+$`)
```
why underscore? Though underscore shouldn't be part of FQDNs and CNs :D

Can you add an example here on what this regex catches and not? Or maybe in unit tests

chrisshino replied:
https://www.digicert.com/kb/ssl-support/underscores-not-allowed-in-fqdns.htm

oh wow, the more you know! Thanks for pointing this out haha I will remove the underscore, honestly I did this a few weeks ago and can't remember why.

rikatz commented on test/e2e/annotations/authtls.go:
```diff
@@ -275,9 +275,9 @@ var _ = framework.DescribeAnnotation("auth-tls-*", func() {
 		annotations := map[string]string{
-			"nginx.ingress.kubernetes.io/auth-tls-secret":        nameSpace + "/" + host,
-			"nginx.ingress.kubernetes.io/auth-tls-verify-client": "on",
-			"nginx.ingress.kubernetes.io/auth-tls-match-cn":      "CN=notgonnamatch",
```
Is this a breaking change? Can users still use CN=something?

rikatz commented on test/e2e/annotations/authtls.go:
```diff
@@ -335,7 +334,7 @@ var _ = framework.DescribeAnnotation("auth-tls-*", func() {
 		annotations := map[string]string{
 			"nginx.ingress.kubernetes.io/auth-tls-secret":        nameSpace + "/" + host,
 			"nginx.ingress.kubernetes.io/auth-tls-verify-client": "on",
-			"nginx.ingress.kubernetes.io/auth-tls-match-cn":      "CN=(itwillmatch|withthenextoption|authtls)",
```
Is removing the CN= requirement a breaking change for this annotation? Or it will still works? I'm worried about removing this and breaking the behavior.

chrisshino replied:
hmm I think this may be tricky, I was asked by my team mate here: https://github.com/kubernetes/ingress-nginx/pull/8434#discussion_r862952023, after the PR was already merged that this "CN=" part is not the intended functionality we were aiming for.

Let me know if you have any ideas, because the way I have it built now its basically grabbing the CN without needing to prefix it with "CN="

chrisshino replied:
Just retested with the "CN=" part added again and everything passes!

No changes will be required for this and will not cause breaks thanks to the !~ (does not regex match), which gives the check that little extra flexibility we want in auth-tls-match-cn.

I still think that for this test though it makes sense to drop the "CN=" part as with the new way that we are capturing the CN directly from ssl_clien_s_dn, it just adds some confusion.

rikatz commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -266,7 +267,9 @@ You can further customize client certificate authentication and behavior with th
 * `nginx.ingress.kubernetes.io/auth-tls-pass-certificate-to-upstream`: Indicates if the received certificates should be passed or not to the upstream server in the header `ssl-client-cert`. Possible values are "true" or "false" (default).
-* `nginx.ingress.kubernetes.io/auth-tls-match-cn`: Adds a sanity check for the CN of the client certificate that is sent over using a string / regex starting with "CN=", example: `"CN=myvalidclient"`. If the certificate CN sent during mTLS does not match your string / regex it will fail with status code 403. Another way of using this is by adding multiple options in your regex, example: `"CN=(opt ...
+* `nginx.ingress.kubernetes.io/auth-tls-match-cn`: Adds a sanity check for the CN of the client certificate that is sent over using a regex. If the certificate CN sent during mTLS does not match your regex, it will fail with status code 403. Another way of using this is by adding multiple options in your regex, example: your actual CN is `"myvalidclient.com"` and the annotation regex you pass in  ...
+* `nginx.ingress.kubernetes.io/auth-tls-exact-match-cn`: Adds a sanity check for the CN of the client certificate that is sent over using a string, example: `"myvalidclient.com"`. If the certificate CN sent during mTLS does not match your string exactly, it will fail with status code 403. With exact match, if you were to pass `"myvalidclient.co"`, it would not match because it is missing the `"m" ...
```
Two insteads in the end of statement (instead you could .... annotation instead)

chrisshino replied:
haha fantastic eye! I will fix this

rikatz commented on rootfs/etc/nginx/template/nginx.tmpl:
```diff
@@ -952,14 +952,33 @@ stream {
 
         set $proxy_upstream_name "-";
 
+        {{ if not ( empty $server.CertificateAuth.ExactMatchCN ) }}
```
We don't add a XOR condition in the annotation parser, saying a user can configure ExactMatchCN OR MatchCN.

This leads me to think this will work, per the template like
* User sets an exactMatchCN
* User then sets a MatchCN
* per Template run, matchCN will replace exactMatchCN

As this is not documented, I would expect that we define a behavior here and use this behavior in the parser, and not in the template. Let me suggest that:

* In the parser, we verify first one annotation, then another. As ExactMatchCN is more restrictive (and new annotation) i would suggest that ExactMatchCN takes precedence over MatchCN, so someone configuring it will know what they are doing.
* We can add in validation webhook a denial saying that this annotations are mutually exclusive if someone tries to add them both. Same on the parser, if both are configured, return an error.
* If you look into the template code, you are going to see that you are repeating the ifs with almost the same information but the name in the struct (ExactMatchCN) and the operation `if ( $ssl_client_s_dn_cn != {{ $server.CertificateAuth.ExactMatchCN }} )` vs `if ( $ssl_client_s_dn_cn !~ {{ $server.CertificateAuth.MatchCN }} ) {`

So, what if you instead of creating a new string annotation, creates a new bool annotation that says if CN should exactly match, and based on this (if true) you change the way your annotation works/get parsed, and the way the template works?

I think this will be the least impact possible on the way it currently works. WDYT?

chrisshino replied:
This makes sense and is something I overlooked. I definitely don't want people using both and having matchCN replace exactMatchCN. I love this idea of adding in the bool annotation, this will work very well I believe!

rikatz reviewed (changes requested):
@chrisshino sorry for the long delay.

As you may know, we are entering in a stabilization period for the code, so I'm not sure we will get this merged before the release cut.

Anyway, I really appreciate all your efforts on this, and really hope you keep contributing.

I've left a comment on one of the reviews on how maybe we should operate here, turning the exact/match behavior as a bool, and mutating the template based on this.

This will turn the operation to be less impactful on the template side, and you can work with the possibilities on the annotation parser.

WDYT?

Also, if you have any questions, please feel free to reach me in Slack.

chrisshino commented:
No worries at all! Thanks for the thorough review once again, I will make the suggested changes! Just let me know what you think regarding my response to your ["breaking changes" question above!](https://github.com/kubernetes/ingress-nginx/pull/8582#discussion_r906866573)

chrisshino commented:
Added the changes requested @rikatz!

I ended up sending this from my personal laptop so its giving me issues with the CLA?

Anyway around this or will I have to redo it on my work laptop?

chrisshino commented:
/assign @rikatz

rikatz commented:
@chrisshino 👋 long time no see, sorry!
So I've left a comment on fixing the exactMatch vs Match, remember? if you implement this, I will gladly approve and merge the PR :)

Thank you!!!

rikatz commented:
/lifecycle active

chrisshino commented:
> @chrisshino 👋 long time no see, sorry!
> 
> So I've left a comment on fixing the exactMatch vs Match, remember? if you implement this, I will gladly approve and merge the PR :)
> 
> 
> 
> Thank you!!!

Hey 👋, I'll take a look tomorrow, I'm pretty sure I addressed all of your comments but I'll try to find where you mentioned that last time and confirm

chrisshino commented:
/assign @rikatz

chrisshino commented:
@rikatz okay so it looks like in all the back and forth I had made a git mistake and pushed the wrong thing, I have since updated that and pushed all my actual fixes that addressed your issues! 

For some reason my CLA is still failing because of a commit message? I dont want to mess with git anymore than I have already lol do you know a good one of fixing this? Then this PR should be good to go! 

Thanks again and sorry it took so long

rikatz commented:
ah yeah, the joys of git :D been there, had the same problem as you, so let's map them:

1) The commit error message is because on one of your commits you added my handler (I saw something like @rikatz comments etc)
2) If by some reason you changed your email during commits, CLA gets kind of crazy. You can even add your other emails on your github profile, or rebase here


To rebase:
- Make a backup of your git folder (in the end, if all goes wrong you can just copy over it! but really, do a backup)
- git remote add upstream git@github.com:kubernetes/ingress-nginx.git
- git rebase upstream/main
   - If there are some conflicts...welll..We can zoom some day so we can look into conflicts and cry together
- If no conflicts, do a "git log". Your commits should be the last ones, over something like upstream/main
- Take the hash of were upstream/main is (like b02130912321c0aa)
- git rebase -i the_hash_above
- A vim/nano will open
- First line you keep as is
- The rest of lines where is written "pick"  you replace for the char "f"
- If you do git log again, now all your commits turned into one
- git push origin branch-name -f

Helps? :D Otherwise, we can see what to do.

chrisshino commented:
@rikatz wow thanks for the step by step instructions!!

Looks like everything should be good now, its just running through CI stuff but I have the correct tags now

sherifkayad commented:
I would like to report a bug in the previous implementation (I am currently using controller version `1.8.1`) and that I would like to kindly see addressed by this PR. Referring to the comment here https://github.com/kubernetes/ingress-nginx/issues/9250#issuecomment-1623506797

Long story short: the combination of using ACME (e.g. letsencrypt) to issue a server cert + using client auth certs on ingress-nginx together with `nginx.ingress.kubernetes.io/auth-tls-match-cn: "CN=my-cn"` would result in a problem that the ACME cert can not be obtained. The reason here is that the CN check also runs for `http` which is wrong in my view and must only be performed when the scheme is `https`. As a workaround / mitigation I stopped using the `nginx.ingress.kubernetes.io/auth-tls-match-cn` annotation and instead I have a server snippet annotation as follows:
```yaml
nginx.ingress.kubernetes.io/server-snippet: |
      set $CERT_CHECK "";
      if ($scheme = https) {
        set $CERT_CHECK "HTTPS";
      }
      if ($ssl_client_s_dn !~ "CN=my-cn") {
        set $CERT_CHECK "${CERT_CHECK}_NOCNMATCH";
      }
      if ($CERT_CHECK = HTTPS_NOCNMATCH) {
        return 403 "client certificate unauthorized: $ssl_client_s_dn\n";
      }
```

In my view, the check using the `auth-tls-match-cn` should be enhanced to be something like the logic I pasted above.
